sdjournal/functions.go

```diff
@@ -0,0 +1,71 @@
+package sdjournal
+
+import (
+	"fmt"
+	"github.com/coreos/pkg/dlopen"
+	"sync"
+	"unsafe"
+)
+
+var (
+	// lazy initialized
+	libsystemdHandle *dlopen.LibHandle
+
+	libsystemdMutex     = &sync.Mutex{}
+	libsystemdFunctions = map[string]unsafe.Pointer{}
+	libsystemdNames     = []string{
+		// systemd < 209
+		"libsystemd-journal.so.0",
+		"libsystemd-journal.so",
+
+		// systemd >= 209 merged libsystemd-journal into libsystemd proper
+		"libsystemd.so.0",
+		"libsystemd.so",
+	}
+)
+
+func getFunction(name string) (unsafe.Pointer, error) {
+	libsystemdMutex.Lock()
+	defer libsystemdMutex.Unlock()
+
+	if libsystemdHandle == nil {
+		handle, err := getHandle()
+		if err != nil {
+			return nil, err
+		}
+
+		libsystemdHandle = handle
+	}
+
+	f, ok := libsystemdFunctions[name]
+	if !ok {
+		var err error
+		f, err = libsystemdHandle.GetSymbolPointer(name)
+		if err != nil {
+			return nil, err
+		}
+
+		libsystemdFunctions[name] = f
+	}
+
+	return f, nil
+}
+
+func getHandle() (*dlopen.LibHandle, error) {
```

Comment:
Removed the deffered at all, since it useless now


On 08/25/2016 03:02 AM, Luca Bruno wrote:
>
> In sdjournal/functions.go 
> <https://github.com/coreos/go-systemd/pull/197#discussion_r76197199>:
>
> > +
> > +	f, ok := libsystemdFunctions[name]
> > +	if !ok {
> > +		var err error
> > +		f, err = libsystemdHandle.GetSymbolPointer(name)
> > +		if err != nil {
> > +			return nil, err
> > +		}
> > +
> > +		libsystemdFunctions[name] = f
> > +	}
> > +
> > +	return f, nil
> > +}
> > +
> > +func getHandle() (*dlopen.LibHandle, error) {
>
> This function (with the deferred cleaner) looks too complex for what 
> it is doing, as it retains some logic from its previous place. I think 
> the |dlopen.GetHandle| can be simply done inside the |getFunction| above.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/coreos/go-systemd/pull/197/files/a90db672d5b1134003c6c70a51ad546fa721f704#r76197199>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ABP5GeQNiM_aumP067mkkM0PqC8X-kZ7ks5qjUwngaJpZM4JsSKO>.
>